crypto/elliptic-curve.ts

```diff
@@ -703,3 +716,369 @@ function createCurveAffine({
+  let y3 = mod(y3Num * y3Den, p);
+
+  return { x: x3, y: y3, infinity: false };
+}
+
+function twistedNegate(
+  { x, y, infinity }: GroupTwisted,
+  p: bigint
+): GroupTwisted {
+  if (infinity) return twistedZero;
+  return { x: x === 0n ? 0n : p - x, y, infinity };
+}
+
+function twistedScale(
+  g: GroupAffine,
+  s: bigint | boolean[],
+  p: bigint,
+  a: bigint,
+  d: bigint
+) {
+  let gProj = projectiveFromTwisted(g);
+  let sgProj = projectiveScaleTwisted(gProj, s, p, a, d);
+  return projectiveToTwisted(sgProj, p);
+}
+
+// https://www.hyperelliptic.org/EFD/g1p/auto-twisted-projective.html
+// https://eprint.iacr.org/2008/013.pdf Section 6
+function projectiveAddTwisted(
+  g: GroupProjective,
+  h: GroupProjective,
+  p: bigint,
+  a: bigint,
+  d: bigint
+): GroupProjective {
+  let { x: X1, y: Y1, z: Z1 } = g;
+  let { x: X2, y: Y2, z: Z2 } = h;
+
+  // A = Z1 * Z2
+  let A = mod(Z1 * Z2, p);
+  // B = A^2
+  let B = mod(A * A, p);
+  // C = X1 * X2
+  let C = mod(X1 * X2, p);
+  // D = Y1 * Y2
+  let D = mod(Y1 * Y2, p);
+  // E = d * C * D
+  let E = mod(d * C * D, p);
+  // F = B - E
+  let F = mod(B - E, p);
+  // G = B + E
+  let G = mod(B + E, p);
+
+  return {
+    x: mod(A * F * ((X1 + Y1) * (X2 + Y2) - C - D), p),
+    y: mod(A * G * (D - a * C), p),
+    z: mod(F * G, p),
+  };
+}
+
+// https://www.hyperelliptic.org/EFD/g1p/auto-twisted-projective.html
+// https://eprint.iacr.org/2008/013.pdf Section 6
+function projectiveDoubleTwisted(
+  g: GroupProjective,
+  p: bigint,
+  a: bigint
+): GroupProjective {
+  let { x: X, y: Y, z: Z } = g;
+
+  // B = (X + Y)^2
+  let B = mod((X + Y) ** 2n, p);
+  // C = X^2
+  let C = mod(X * X, p);
+  // D = Y^2
+  let D = mod(Y * Y, p);
+  // E = a * C
+  let E = mod(a * C, p);
+  // F = E + D
+  let F = mod(E + D, p);
+  // H = Z^2
+  let H = mod(Z * Z, p);
+  // J =  F - 2 * H
+  let J = mod(F - 2n * H, p);
+
+  return {
+    x: mod((B - C - D) * J, p),
+    y: mod(F * (E - D), p),
+    z: mod(F * J, p),
+  };
+}
+
+function projectiveScaleTwisted(
+  g: GroupProjective,
+  x: bigint | boolean[],
+  p: bigint,
+  a: bigint,
+  d: bigint
+) {
+  let bits = typeof x === 'bigint' ? bigIntToBits(x) : x;
+  let h = projectiveZeroTwisted;
+  for (let bit of bits) {
+    if (bit) h = projectiveAddTwisted(h, g, p, a, d);
+    g = projectiveDoubleTwisted(g, p, a);
+  }
+  return h;
+}
+
+function projectiveFromTwisted({
+  x,
+  y,
+  infinity,
+}: GroupTwisted): GroupProjective {
+  if (infinity) return projectiveZeroTwisted;
+  return { x, y, z: 1n };
+}
+
+// The twisted curve with equation
+// a * x^2 + y^2 = 1 + d * x^2 * y^2
+// in projective coordinates is represented as
+// a * X^2 * Z^2 + Y^2 Z^2 = Z^4 + d * X^2 * Y^2
+// where x = X/Z, y = Y/Z, and Z ≠ 0
+function projectiveToTwisted(g: GroupProjective, p: bigint): GroupTwisted {
+  let z = g.z;
+  if (z === 0n) {
+    // degenerate case
+    return twistedZero;
+  } else if (z === 1n && g.x === 0n && g.y === 1n) {
```

Comment:
If I didn't treat this case separately, the following `else if` would place a `false` in the `infinity` parameter and then unit tests would complain because they used to find assertions of (0,1,false) != (0,1,true). But note that this is a branch that does not exist in the affine counterpart, and it is assuming that the "only" infinity point is exactly (0,1,1).